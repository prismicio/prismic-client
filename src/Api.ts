import Predicates from './Predicates';
import { RequestHandler, RequestCallback } from './request';
import { Document } from './documents';
import { ApiCache } from './cache';
import ResolvedApi, { ApiData } from './ResolvedApi';
import HttpClient from './HttpClient';

export interface ApiOptions {
  accessToken?: string;
  requestHandler?: RequestHandler;
  req?: any;
  apiCache?: ApiCache;
  apiDataTTL?: number;
  proxyAgent?: any;
}

export default class PrismicApi {
  url: string;
  options: ApiOptions;
  apiDataTTL: number;
  httpClient: HttpClient;

  constructor(url: string, options?: ApiOptions) {
    this.options = options || {};
    this.url = url;
    if (this.options.accessToken) {
      const accessTokenParam = `access_token=${this.options.accessToken}`;
      this.url += (url.indexOf('?') > -1 ? '&' : '?') + accessTokenParam;
    }
    this.apiDataTTL = this.options.apiDataTTL || 5;
    this.httpClient = new HttpClient(
      this.options.requestHandler,
      this.options.apiCache,
      this.options.proxyAgent,
    );
  }

  /**
   * Fetches data used to construct the api client, from cache if it's
   * present, otherwise from calling the prismic api endpoint (which is
   * then cached).
   */
  get(cb?: RequestCallback<ResolvedApi>): Promise<ResolvedApi> {
    return this.httpClient.cachedRequest<ApiData>(this.url, { ttl: this.apiDataTTL }).then((data) => {
      const resolvedApi = new ResolvedApi(data, this.httpClient, this.options);
      cb && cb(null, resolvedApi);
      return resolvedApi;
    }).catch((error) => {
      cb && cb(error);
      throw error;
    });
  }
}
