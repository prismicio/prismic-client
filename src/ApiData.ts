export interface QuickRoutes {
  url: string;
  lastUpdate: number;
  enabled: boolean;
}

export interface ApiData {
  refs: Ref[];
  bookmarks: { [key: string]: string };
  types: { [key: string]: string };
  tags: string[];
  forms: { [key: string]: Form };
  quickRoutes: QuickRoutes;
  experiments: any;
  oauth_initiate: string;
  oauth_token: string;
  version: string;
  licence: string;
}
