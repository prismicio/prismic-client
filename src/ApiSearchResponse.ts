import { Document } from "./documents";

export default interface ApiSearchResponse<T = any> {
  page: number;
  results_per_page: number;
  results_size: number;
  total_results_size: number;
  total_pages: number;
  next_page: string;
  prev_page: string;
  results: Document<T>[];
}
