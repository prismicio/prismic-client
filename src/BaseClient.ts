import { type LimitFunction, pLimit } from "./lib/pLimit"

import { PrismicError } from "./errors/PrismicError"

/**
 * The default delay used with APIs not providing rate limit headers.
 *
 * @internal
 */
export const UNKNOWN_RATE_LIMIT_DELAY = 1500

/**
 * A universal API to make network requests. A subset of the `fetch()` API.
 *
 * {@link https://developer.mozilla.org/en-US/docs/Web/API/fetch}
 */
export type FetchLike = (
	input: string,
	init?: RequestInitLike,
) => Promise<ResponseLike>

/**
 * An object that allows you to abort a `fetch()` request if needed via an
 * `AbortController` object
 *
 * {@link https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal}
 */
// `any` is used often here to ensure this type is universally valid among
// different AbortSignal implementations. The types of each property are not
// important to validate since it is blindly passed to a given `fetch()`
// function.
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type AbortSignalLike = any

/**
 * A subset of RequestInit properties to configure a `fetch()` request.
 */
// Only options relevant to the client are included. Extending from the full
// RequestInit would cause issues, such as accepting Header objects.
//
// An interface is used to allow other libraries to augment the type with
// environment-specific types.
export interface RequestInitLike extends Pick<RequestInit, "cache"> {
	/**
	 * The HTTP method to use for the request.
	 */
	method?: string

	/**
	 * The request body to send with the request.
	 */
	// We want to keep the body type as compatible as possible, so
	// we only declare the type we need and accept anything else.
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	body?: any | FormData | string

	/**
	 * An object literal to set the `fetch()` request's headers.
	 */
	headers?: Record<string, string>

	/**
	 * An AbortSignal to set the `fetch()` request's signal.
	 *
	 * See:
	 * [https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal)
	 */
	// NOTE: `AbortSignalLike` is `any`! It is left as `AbortSignalLike`
	// for backwards compatibility (the type is exported) and to signal to
	// other readers that this should be an AbortSignal-like object.
	signal?: AbortSignalLike
}

/**
 * The minimum required properties from Response.
 */
export interface ResponseLike {
	ok: boolean
	status: number
	headers: HeadersLike
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	json(): Promise<any>
	text(): Promise<string>
	blob(): Promise<Blob>
}

/**
 * The minimum required properties from Headers.
 */
export interface HeadersLike {
	get(name: string): string | null
}

/**
 * The minimum required properties to treat as an HTTP Request for automatic
 * Prismic preview support.
 */
export type HttpRequestLike =
	| /**
	 * Web API Request
	 *
	 * @see http://developer.mozilla.org/en-US/docs/Web/API/Request
	 */
	{
			headers?: {
				get(name: string): string | null
			}
			url?: string
	  }

	/**
	 * Express-style Request
	 */
	| {
			headers?: {
				cookie?: string
			}
			query?: Record<string, unknown>
	  }

/**
 * Configuration for clients that determine how APIs are queried.
 */
export type BaseClientConfig = {
	/**
	 * The function used to make network requests to the Prismic REST API. In
	 * environments where a global `fetch` function does not exist, such as
	 * Node.js, this function must be provided.
	 */
	fetch?: FetchLike

	/**
	 * Options provided to the client's `fetch()` on all network requests. These
	 * options will be merged with internally required options. They can also be
	 * overriden on a per-query basis using the query's `fetchOptions` parameter.
	 */
	fetchOptions?: RequestInitLike
}

/**
 * Parameters for any client method that use `fetch()`.
 */
export type FetchParams = {
	/**
	 * Options provided to the client's `fetch()` on all network requests. These
	 * options will be merged with internally required options. They can also be
	 * overriden on a per-query basis using the query's `fetchOptions` parameter.
	 */
	fetchOptions?: RequestInitLike

	/**
	 * An `AbortSignal` provided by an `AbortController`. This allows the network
	 * request to be cancelled if necessary.
	 *
	 * @deprecated Move the `signal` parameter into `fetchOptions.signal`:
	 *
	 * @see \<https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal\>
	 */
	signal?: AbortSignalLike
}

/**
 * The result of a `fetch()` job.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type FetchJobResult<TJSON = any> = {
	status: number
	headers: HeadersLike
	json: TJSON
	text?: string
}

export abstract class BaseClient {
	/**
	 * The function used to make network requests to the Prismic REST API. In
	 * environments where a global `fetch` function does not exist, such as
	 * Node.js, this function must be provided.
	 */
	fetchFn: FetchLike

	fetchOptions?: RequestInitLike

	/**
	 * Active queued `fetch()` jobs keyed by URL and AbortSignal (if it exists).
	 */
	private queuedFetchJobs: Record<string, LimitFunction> = {}

	/**
	 * Active deduped `fetch()` jobs keyed by URL and AbortSignal (if it exists).
	 */
	private dedupedFetchJobs: Record<
		string,
		Map<AbortSignalLike | undefined, Promise<FetchJobResult>>
	> = {}

	constructor(options: BaseClientConfig) {
		this.fetchOptions = options.fetchOptions

		if (typeof options.fetch === "function") {
			this.fetchFn = options.fetch
		} else if (typeof globalThis.fetch === "function") {
			this.fetchFn = globalThis.fetch as FetchLike
		} else {
			throw new PrismicError(
				"A valid fetch implementation was not provided. In environments where fetch is not available (including Node.js), a fetch implementation must be provided via a polyfill or the `fetch` option.",
				undefined,
				undefined,
			)
		}

		// If the global fetch function is used, we must bind it to the global scope.
		if (this.fetchFn === globalThis.fetch) {
			this.fetchFn = this.fetchFn.bind(globalThis)
		}
	}

	protected async fetch(
		url: string,
		params: FetchParams = {},
	): Promise<FetchJobResult> {
		const requestInit: RequestInitLike = {
			...this.fetchOptions,
			...params.fetchOptions,
			headers: {
				...this.fetchOptions?.headers,
				...params.fetchOptions?.headers,
			},
			signal:
				params.fetchOptions?.signal ||
				params.signal ||
				this.fetchOptions?.signal,
		}

		// Request with a `body` are throttled, others are deduped.
		if (params.fetchOptions?.body) {
			return this.queueFetch(url, requestInit)
		} else {
			return this.dedupeFetch(url, requestInit)
		}
	}

	private queueFetch(
		url: string,
		requestInit: RequestInitLike = {},
	): Promise<FetchJobResult> {
		// Rate limiting is done per hostname.
		const hostname = new URL(url).hostname

		if (!this.queuedFetchJobs[hostname]) {
			this.queuedFetchJobs[hostname] = pLimit({
				interval: UNKNOWN_RATE_LIMIT_DELAY,
			})
		}

		return this.queuedFetchJobs[hostname](() =>
			this.createFetchJob(url, requestInit),
		)
	}

	private dedupeFetch(
		url: string,
		requestInit: RequestInitLike = {},
	): Promise<FetchJobResult> {
		let job: Promise<FetchJobResult>

		// `fetchJobs` is keyed twice: first by the URL and again by is
		// signal, if one exists.
		//
		// Using two keys allows us to reuse fetch requests for
		// equivalent URLs, but eject when we detect unique signals.
		if (
			this.dedupedFetchJobs[url] &&
			this.dedupedFetchJobs[url].has(requestInit.signal)
		) {
			// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
			job = this.dedupedFetchJobs[url].get(requestInit.signal)!
		} else {
			this.dedupedFetchJobs[url] = this.dedupedFetchJobs[url] || new Map()

			job = this.createFetchJob(url, requestInit).finally(() => {
				this.dedupedFetchJobs[url]?.delete(requestInit.signal)

				if (this.dedupedFetchJobs[url]?.size === 0) {
					delete this.dedupedFetchJobs[url]
				}
			})

			this.dedupedFetchJobs[url].set(requestInit.signal, job)
		}

		return job
	}

	private createFetchJob(
		url: string,
		requestInit: RequestInitLike = {},
	): Promise<FetchJobResult> {
		return this.fetchFn(url, requestInit).then(async (res) => {
			// We can assume Prismic REST API responses
			// will have a `application/json`
			// Content Type. If not, this will
			// throw, signaling an invalid
			// response.
			// eslint-disable-next-line @typescript-eslint/no-explicit-any
			let json: any = undefined
			let text: string | undefined = undefined
			if (res.ok) {
				try {
					json = await res.json()
				} catch {
					// noop
				}
			} else {
				try {
					text = await res.text()
					json = JSON.parse(text)
				} catch {
					// noop
				}
			}

			return {
				status: res.status,
				headers: res.headers,
				json,
				text,
			}
		})
	}
}
