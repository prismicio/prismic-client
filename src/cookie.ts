/**
 * The well-known name of the cookie used to store a Prismic preview session's
 * ref.
 */
export const preview = "io.prismic.preview"
