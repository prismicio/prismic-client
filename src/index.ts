// Primary library API.
import { getRepositoryEndpoint } from "./getRepositoryEndpoint";
export { getRepositoryEndpoint };
/**
 * @deprecated Renamed to `getRepositoryEndpoint`.
 */
// TODO: Remove in v3.
export const getEndpoint = getRepositoryEndpoint;
export { getRepositoryName } from "./getRepositoryName";
export { getGraphQLEndpoint } from "./getGraphQLEndpoint";
export { isRepositoryName } from "./isRepositoryName";
export { isRepositoryEndpoint } from "./isRepositoryEndpoint";
export { buildQueryURL } from "./buildQueryURL";
export { createClient, Client } from "./client";

// Predicates API.
import { predicate } from "./predicate";
export { predicate };
/**
 * @deprecated Renamed to `predicate` (without an "s").
 */
// TODO: Remove in v3.
export const predicates = predicate;
/**
 * @deprecated Renamed to `predicate` (lowercase and without an "s").
 */
// TODO: Remove in v3.
export const Predicates = predicate;

// Custom errors used by Client.
export { PrismicError } from "./PrismicError";
export { ForbiddenError } from "./ForbiddenError";
export { ParsingError } from "./ParsingError";
export { NotFoundError } from "./NotFoundError";

// A collection of well-known cookie names shared between Prismic libraries and systems.
export * as cookie from "./cookie";

// General types used throughout the project. These are made public to allow users to better type their projects.
export type { CreateClient, ClientConfig } from "./client";
export type { QueryParams, BuildQueryURLArgs } from "./buildQueryURL";
export type {
	AbortSignalLike,
	FetchLike,
	HttpRequestLike,
	Ordering,
	RequestInitLike,
	ResponseLike,
	Route,
} from "./types";
