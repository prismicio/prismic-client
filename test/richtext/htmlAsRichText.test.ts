import { describe, expect, it, vi } from "vitest";

import { LinkType, asHTML } from "../../src";
import {
	AsRichTextConfig,
	htmlAsRichText,
	htmlAsRichTextSync,
} from "../../src/richtext";

type Case = {
	case: string;

	input: string;

	config?: AsRichTextConfig;

	/**
	 * The rich text format is a lossy representation of HTML. Namely it does not
	 * preserves indentation and applies some optimizations to the output such as
	 * merging directly adjacent identical spans.
	 *
	 * By default, the test suite will expect the HTML representation of the
	 * output to exactly match the input. This flag can be used to tell it to
	 * expect the output to not match the input instead.
	 */
	expectAsHTMLNotToMatchInput?: true;
};

describe.each<Case>([
	{
		case: "empty",
		input: /* html */ ``,
	},
	{
		case: "single tag",
		input: /* html */ `<p>lorem ipsum dolor sit amet</p>`,
	},
	{
		case: "multiple tags",
		input: /* html */ `<h1>lorem ipsum dolor sit amet</h1><p>consectetur adipiscing elit</p>`,
	},
	{
		case: "spans (strong, em)",
		input: /* html */ `<p>lorem <strong>ipsum</strong> dolor <em>sit</em> amet</p>`,
	},
	{
		case: "spans (label)",
		input: /* html */ `<p>lorem <span class="underline">ipsum</span> dolor sit amet</p>`,
		config: { converter: { "span.underline": { label: "underline" } } },
	},
	{
		case: "spans (hyperlink)",
		input: /* html */ `<p>lorem <a href="https://prismic.io">ipsum</a> dolor <a href="https://prismic.io" target="_blank">sit</a> amet</p>`,
	},
	{
		case: "nested spans",
		input: /* html */ `<p>lorem <strong>ips<em>um <a href="https://prismic.io">dolor</a></em></strong> sit amet</p>`,
	},
	{
		case: "compacts directly adjacent identical spans (strong, em)",
		input: /* html */ `<p>lorem <strong>ipsum</strong><strong> dolor</strong> sit <em>amet</em><em> consectetur</em> adipiscing elit</p>`,
		// `strong` and `em` tags will be merged into single ones.
		expectAsHTMLNotToMatchInput: true,
	},
	{
		case: "compacts directly adjacent identical spans (hyperlink)",
		input: /* html */ `<p>lorem <a href="https://prismic.io">ipsum</a><a href="https://prismic.io"> dolor</a> sit amet</p>`,
		// `a` tags will be merged into single ones.
		expectAsHTMLNotToMatchInput: true,
	},
	{
		case: "compacts directly adjacent identical spans (label)",
		input: /* html */ `<p>lorem <span class="underline">ipsum</span><span class="underline"> dolor</span> sit amet</p>`,
		config: { converter: { "span.underline": { label: "underline" } } },
		// `span.underline` tags will be merged into single ones.
		expectAsHTMLNotToMatchInput: true,
	},
	{
		case: "does not compact directly adjacent different spans (hyperlink)",
		input: /* html */ `<p>lorem <a href="https://prismic.io">ipsum</a><a href="https://google.com"> dolor</a> sit amet</p>`,
	},
	{
		case: "does not compact directly adjacent different spans (label)",
		input: /* html */ `<p>lorem <span class="underline">ipsum</span><span class="strikethrough"> dolor</span> sit amet</p>`,
		config: {
			converter: {
				"span.underline": { label: "underline" },
				"span.strikethrough": { label: "strikethrough" },
			},
		},
	},
	{
		case: "compacts nested directly adjacent identical spans",
		input: /* html */ `<p>lorem <strong>ips<em>um</em></strong><em> dolor</em> sit amet</p>`,
	},
	{
		case: "image",
		input: /* html */ `<img src="https://example.com/foo.png" alt="foo" />`,
	},
	{
		case: "image (prismic)",
		input: /* html */ `<img src="https://images.prismic.io/200629-sms-hoy/f0a757f6-770d-4eb8-a08b-f1727f1a58e4_guilherme-romano-KI2KaOeT670-unsplash.jpg?auto=format%2Ccompress&rect=399%2C259%2C1600%2C1068&w=2400&h=1602" alt="foo" />`,
	},
	{
		case: "embed",
		input: /* html */ `<iframe width="200" height="150" src="https://www.youtube.com/embed/wkS1bf7BLjs?feature=oembed" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen title="幾田りら「ハミング」Official Music Video"></iframe>`,
	},
	{
		case: "extracts images from rich text text nodes and resumes with the same text node",
		input: /* html */ `<p>lorem ipsum <img src="https://example.com/foo.png" alt="bar" /> dolor sit amet</p>`,
		expectAsHTMLNotToMatchInput: true,
	},
	{
		case: "extracts images from rich text text nodes and resumes with the same text node (spans)",
		input: /* html */ `<p><strong>lorem</strong> ipsum <img src="https://example.com/foo.png" alt="bar" /> dolor <em>sit</em> amet</p>`,
		expectAsHTMLNotToMatchInput: true,
	},
	{
		case: "extracts images from rich text text nodes and resumes with the same text node (adjacent spans)",
		input: /* html */ `<p>lorem <strong>ipsum</strong> <img src="https://example.com/foo.png" alt="bar" /> <em>dolor</em> sit amet</p>`,
		expectAsHTMLNotToMatchInput: true,
	},
	{
		case: "converts only the given container",
		input: /* html */ `
			<article id="foo"><p>lorem ipsum dolor sit amet</p></article>
			<article id="bar"><p>consectetur adipiscing elit</p></article>`,
		config: { container: "article#bar" },
		expectAsHTMLNotToMatchInput: true,
	},
	{
		case: "excludes the given selectors",
		input: /* html */ `
			<h1><a href="#">lorem</a> ipsum dolor sit amet</h1>
			<p><a href="#">consectetur</a> adipiscing elit</p>`,
		config: { exclude: ["h1 > a"] },
		expectAsHTMLNotToMatchInput: true,
	},
	{
		case: "treats `<br>` as new lines",
		input: /* html */ `<p>lorem ipsum dolor sit amet<br />consectetur adipiscing elit</p>`,
	},
	{
		case: "strips indentation",
		input: /* html */ `
			<p>
				lorem ipsum dolor sit amet
			</p>
		`,
		expectAsHTMLNotToMatchInput: true,
	},
	{
		case: "strips complex indentation",
		input: /* html */ `
			<p>
				lorem ipsum dolor sit amet
        consectetur adipiscing elit
			</p>
		`,
		expectAsHTMLNotToMatchInput: true,
	},
])(
	"transforms HTML to rich text ($case)",
	({ input, config, expectAsHTMLNotToMatchInput }) => {
		it("produces the same output with sync and async helpers", async () => {
			const asyncOutput = await htmlAsRichText(input, config);
			const syncOutput = htmlAsRichTextSync(input, config);

			expect(asyncOutput).toStrictEqual(syncOutput);
		});

		it("produces valid rich text field", async () => {
			const output = await htmlAsRichText(input, config);

			expect(output.result).toMatchSnapshot();

			const outputAsHTML = asHTML(output.result, {
				serializer: {
					// A simplified hyperlink serializer so that we don't have
					// to append `rel="noopener noreferrer"` to every link in
					// the test cases.
					hyperlink: ({ node, children }) => {
						const maybeTarget =
							node.data.link_type === LinkType.Web && node.data.target
								? ` target="${node.data.target}"`
								: "";

						return `<a href="${node.data.url}"${maybeTarget}>${children}</a>`;
					},
					// A simplified image serializer so that we don't have to
					// wrap the images in a `div` element like the default
					// serializer does.
					image: ({ node }) => `<img src="${node.url}" alt="${node.alt}" />`,
					// A simplified embed serializer so that we don't have to
					// wrap the embeds in a `div` element with the various data
					// attributes the default serializer applies.
					embed: ({ node }) => node.oembed.html,
				},
			});

			if (!expectAsHTMLNotToMatchInput) {
				expect(outputAsHTML).toBe(input);
			} else {
				expect(outputAsHTML).not.toBe(input);
			}
		});
	},
);

it("warns about missing `src` attribute in `image` elements", async () => {
	const consoleWarnSpy = vi
		.spyOn(console, "warn")
		.mockImplementation(() => void 0);

	await htmlAsRichText(/* html */ `<img>`);

	expect(consoleWarnSpy).toHaveBeenCalledWith(
		expect.stringMatching(/missing-image-src/i),
	);

	consoleWarnSpy.mockRestore();
});

it("warns about missing `src` attribute in `embed` elements", async () => {
	const consoleWarnSpy = vi
		.spyOn(console, "warn")
		.mockImplementation(() => void 0);

	await htmlAsRichText(
		/* html */ `<iframe>lorem ipsum dolor sit amet</iframe>`,
	);

	expect(consoleWarnSpy).toHaveBeenCalledWith(
		expect.stringMatching(/missing-embed-src/i),
	);

	consoleWarnSpy.mockRestore();
});

it("warns about missing `href` attribute in `hyperlink` elements", async () => {
	const consoleWarnSpy = vi
		.spyOn(console, "warn")
		.mockImplementation(() => void 0);

	await htmlAsRichText(/* html */ `<p><a>missing-hyperlink-href</a></p>`);

	expect(consoleWarnSpy).toHaveBeenCalledWith(
		expect.stringMatching(/missing-hyperlink-href/i),
	);

	consoleWarnSpy.mockRestore();
});
