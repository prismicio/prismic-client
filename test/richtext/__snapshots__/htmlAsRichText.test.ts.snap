// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transforms HTML to rich text ('compacts directly adjacent identical spans (hyperlink)') > produces valid rich text field 1`] = `
[
  {
    "spans": [
      {
        "data": {
          "link_type": "Web",
          "target": undefined,
          "url": "https://prismic.io",
        },
        "end": 17,
        "start": 6,
        "type": "hyperlink",
      },
    ],
    "text": "lorem ipsum dolor sit amet",
    "type": "paragraph",
  },
]
`;

exports[`transforms HTML to rich text ('compacts directly adjacent identical spans (label)') > produces valid rich text field 1`] = `
[
  {
    "spans": [
      {
        "data": {
          "label": "underline",
        },
        "end": 17,
        "start": 6,
        "type": "label",
      },
    ],
    "text": "lorem ipsum dolor sit amet",
    "type": "paragraph",
  },
]
`;

exports[`transforms HTML to rich text ('compacts directly adjacent identical spans (strong, em)') > produces valid rich text field 1`] = `
[
  {
    "spans": [
      {
        "end": 17,
        "start": 6,
        "type": "strong",
      },
      {
        "end": 38,
        "start": 22,
        "type": "em",
      },
    ],
    "text": "lorem ipsum dolor sit amet consectetur adipiscing elit",
    "type": "paragraph",
  },
]
`;

exports[`transforms HTML to rich text ('compacts nested directly adjacent identical spans') > produces valid rich text field 1`] = `
[
  {
    "spans": [
      {
        "end": 11,
        "start": 6,
        "type": "strong",
      },
      {
        "end": 17,
        "start": 9,
        "type": "em",
      },
    ],
    "text": "lorem ipsum dolor sit amet",
    "type": "paragraph",
  },
]
`;

exports[`transforms HTML to rich text ('converts only the given container') > produces valid rich text field 1`] = `
[
  {
    "spans": [],
    "text": "consectetur adipiscing elit",
    "type": "paragraph",
  },
]
`;

exports[`transforms HTML to rich text ('does not compact directly adjacent different spans (hyperlink)') > produces valid rich text field 1`] = `
[
  {
    "spans": [
      {
        "data": {
          "link_type": "Web",
          "target": undefined,
          "url": "https://prismic.io",
        },
        "end": 11,
        "start": 6,
        "type": "hyperlink",
      },
      {
        "data": {
          "link_type": "Web",
          "target": undefined,
          "url": "https://google.com",
        },
        "end": 17,
        "start": 11,
        "type": "hyperlink",
      },
    ],
    "text": "lorem ipsum dolor sit amet",
    "type": "paragraph",
  },
]
`;

exports[`transforms HTML to rich text ('does not compact directly adjacent different spans (label)') > produces valid rich text field 1`] = `
[
  {
    "spans": [
      {
        "data": {
          "label": "underline",
        },
        "end": 11,
        "start": 6,
        "type": "label",
      },
      {
        "data": {
          "label": "strikethrough",
        },
        "end": 17,
        "start": 11,
        "type": "label",
      },
    ],
    "text": "lorem ipsum dolor sit amet",
    "type": "paragraph",
  },
]
`;

exports[`transforms HTML to rich text ('embed') > produces valid rich text field 1`] = `
[
  {
    "oembed": {
      "embed_url": "https://www.youtube.com/embed/wkS1bf7BLjs?feature=oembed",
      "height": 150,
      "html": "<iframe width=\\"200\\" height=\\"150\\" src=\\"https://www.youtube.com/embed/wkS1bf7BLjs?feature=oembed\\" frameborder=\\"0\\" allow=\\"accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share\\" referrerpolicy=\\"strict-origin-when-cross-origin\\" allowfullscreen title=\\"幾田りら「ハミング」Official Music Video\\"></iframe>",
      "title": "幾田りら「ハミング」Official Music Video",
      "type": "rich",
      "version": "1.0",
      "width": 200,
    },
    "type": "embed",
  },
]
`;

exports[`transforms HTML to rich text ('empty') > produces valid rich text field 1`] = `[]`;

exports[`transforms HTML to rich text ('excludes the given selectors') > produces valid rich text field 1`] = `
[
  {
    "spans": [],
    "text": "ipsum dolor sit amet",
    "type": "heading1",
  },
  {
    "spans": [
      {
        "data": {
          "link_type": "Web",
          "target": undefined,
          "url": "#",
        },
        "end": 11,
        "start": 0,
        "type": "hyperlink",
      },
    ],
    "text": "consectetur adipiscing elit",
    "type": "paragraph",
  },
]
`;

exports[`transforms HTML to rich text ('extracts images from rich text text nodes and resumes with the same text node (adjacent spans)') > produces valid rich text field 1`] = `
[
  {
    "spans": [
      {
        "end": 11,
        "start": 6,
        "type": "strong",
      },
    ],
    "text": "lorem ipsum",
    "type": "paragraph",
  },
  {
    "alt": "bar",
    "copyright": null,
    "dimensions": {
      "height": 0,
      "width": 0,
    },
    "edit": {
      "background": "transparent",
      "x": 0,
      "y": 0,
      "zoom": 1,
    },
    "id": "",
    "type": "image",
    "url": "https://example.com/foo.png",
  },
  {
    "spans": [
      {
        "end": 5,
        "start": 0,
        "type": "em",
      },
    ],
    "text": "dolor sit amet",
    "type": "paragraph",
  },
]
`;

exports[`transforms HTML to rich text ('extracts images from rich text text nodes and resumes with the same text node (spans)') > produces valid rich text field 1`] = `
[
  {
    "spans": [
      {
        "end": 5,
        "start": 0,
        "type": "strong",
      },
    ],
    "text": "lorem ipsum",
    "type": "paragraph",
  },
  {
    "alt": "bar",
    "copyright": null,
    "dimensions": {
      "height": 0,
      "width": 0,
    },
    "edit": {
      "background": "transparent",
      "x": 0,
      "y": 0,
      "zoom": 1,
    },
    "id": "",
    "type": "image",
    "url": "https://example.com/foo.png",
  },
  {
    "spans": [
      {
        "end": 9,
        "start": 6,
        "type": "em",
      },
    ],
    "text": "dolor sit amet",
    "type": "paragraph",
  },
]
`;

exports[`transforms HTML to rich text ('extracts images from rich text text nodes and resumes with the same text node') > produces valid rich text field 1`] = `
[
  {
    "spans": [],
    "text": "lorem ipsum",
    "type": "paragraph",
  },
  {
    "alt": "bar",
    "copyright": null,
    "dimensions": {
      "height": 0,
      "width": 0,
    },
    "edit": {
      "background": "transparent",
      "x": 0,
      "y": 0,
      "zoom": 1,
    },
    "id": "",
    "type": "image",
    "url": "https://example.com/foo.png",
  },
  {
    "spans": [],
    "text": "dolor sit amet",
    "type": "paragraph",
  },
]
`;

exports[`transforms HTML to rich text ('image (prismic)') > produces valid rich text field 1`] = `
[
  {
    "alt": "foo",
    "copyright": null,
    "dimensions": {
      "height": 1602,
      "width": 2400,
    },
    "edit": {
      "background": "transparent",
      "x": 399,
      "y": 259,
      "zoom": 1.5,
    },
    "id": "",
    "type": "image",
    "url": "https://images.prismic.io/200629-sms-hoy/f0a757f6-770d-4eb8-a08b-f1727f1a58e4_guilherme-romano-KI2KaOeT670-unsplash.jpg?auto=format%2Ccompress&rect=399%2C259%2C1600%2C1068&w=2400&h=1602",
  },
]
`;

exports[`transforms HTML to rich text ('image') > produces valid rich text field 1`] = `
[
  {
    "alt": "foo",
    "copyright": null,
    "dimensions": {
      "height": 0,
      "width": 0,
    },
    "edit": {
      "background": "transparent",
      "x": 0,
      "y": 0,
      "zoom": 1,
    },
    "id": "",
    "type": "image",
    "url": "https://example.com/foo.png",
  },
]
`;

exports[`transforms HTML to rich text ('multiple tags') > produces valid rich text field 1`] = `
[
  {
    "spans": [],
    "text": "lorem ipsum dolor sit amet",
    "type": "heading1",
  },
  {
    "spans": [],
    "text": "consectetur adipiscing elit",
    "type": "paragraph",
  },
]
`;

exports[`transforms HTML to rich text ('nested spans') > produces valid rich text field 1`] = `
[
  {
    "spans": [
      {
        "end": 17,
        "start": 6,
        "type": "strong",
      },
      {
        "end": 17,
        "start": 9,
        "type": "em",
      },
      {
        "data": {
          "link_type": "Web",
          "target": undefined,
          "url": "https://prismic.io",
        },
        "end": 17,
        "start": 12,
        "type": "hyperlink",
      },
    ],
    "text": "lorem ipsum dolor sit amet",
    "type": "paragraph",
  },
]
`;

exports[`transforms HTML to rich text ('single tag') > produces valid rich text field 1`] = `
[
  {
    "spans": [],
    "text": "lorem ipsum dolor sit amet",
    "type": "paragraph",
  },
]
`;

exports[`transforms HTML to rich text ('spans (hyperlink)') > produces valid rich text field 1`] = `
[
  {
    "spans": [
      {
        "data": {
          "link_type": "Web",
          "target": undefined,
          "url": "https://prismic.io",
        },
        "end": 11,
        "start": 6,
        "type": "hyperlink",
      },
      {
        "data": {
          "link_type": "Web",
          "target": "_blank",
          "url": "https://prismic.io",
        },
        "end": 21,
        "start": 18,
        "type": "hyperlink",
      },
    ],
    "text": "lorem ipsum dolor sit amet",
    "type": "paragraph",
  },
]
`;

exports[`transforms HTML to rich text ('spans (label)') > produces valid rich text field 1`] = `
[
  {
    "spans": [
      {
        "data": {
          "label": "underline",
        },
        "end": 11,
        "start": 6,
        "type": "label",
      },
    ],
    "text": "lorem ipsum dolor sit amet",
    "type": "paragraph",
  },
]
`;

exports[`transforms HTML to rich text ('spans (strong, em)') > produces valid rich text field 1`] = `
[
  {
    "spans": [
      {
        "end": 11,
        "start": 6,
        "type": "strong",
      },
      {
        "end": 21,
        "start": 18,
        "type": "em",
      },
    ],
    "text": "lorem ipsum dolor sit amet",
    "type": "paragraph",
  },
]
`;

exports[`transforms HTML to rich text ('strips complex indentation') > produces valid rich text field 1`] = `
[
  {
    "spans": [],
    "text": "lorem ipsum dolor sit amet consectetur adipiscing elit",
    "type": "paragraph",
  },
]
`;

exports[`transforms HTML to rich text ('strips indentation') > produces valid rich text field 1`] = `
[
  {
    "spans": [],
    "text": "lorem ipsum dolor sit amet",
    "type": "paragraph",
  },
]
`;

exports[`transforms HTML to rich text ('treats \`<br>\` as new lines') > produces valid rich text field 1`] = `
[
  {
    "spans": [],
    "text": "lorem ipsum dolor sit amet
consectetur adipiscing elit",
    "type": "paragraph",
  },
]
`;
